Add latest changes by parsing closed PRs and then filtering

mebenn opened the pull request:
Please do not merge without editing.

I noticed that  'latest_changes.md' hadn't been updated for a while, so I used the github Issues API to get a list of all closed Pull Requests from OpenIndiana/oi-userland since 2019, filter the 'title' and 'closed_at' tags, dumped these into a script so that I could print these as required by the markup in latest_changes.md, and, finally, I dumped the output into latest_changes.md. 

There appears to be too much detail that requires removal. I don't think _all_ Pull Requests should be _explicitely_ included in our release notes. Any ideas how this might be better handled in the future? While all PRs should be included in future releases, not all PRs should necessarily be mentioned in the release notes as the release notes would be exhaustively long. If we could set one of the tags that github returns for the PR, that would be good, e.g., milestone always appears to be null.
Is there a convenient way to group PRs, i.e., bumped version of some app.

Moreover, I don't pretend to understand the API, I can't get the 'since' filter working properly; so I cannot exclude errors originating from me.

We could easily generate latest_changes.md entirely by script, if only I could properly understand the issues API.

AndWac commented on docs/release-notes/latest-changes.md:
```diff
@@ -28,6 +28,56 @@ All Rights Reserved. (Contributor contact(s):________________[Insert hyperlink/a
 Date       | Change
 ---------  | -------------
+2022-02-03 | fix terminator history
+2022-02-04 | barman: update to 2.18
```
fix terminator history was necessary to obsolete the python-3.5 version of terminator. This entry should be removed.

mebenn replied:
Done

AndWac commented on docs/release-notes/latest-changes.md:
```diff
@@ -28,6 +28,56 @@ All Rights Reserved. (Contributor contact(s):________________[Insert hyperlink/a
+2022-02-01 | fonts/terminus: update to 4.49.1
+2022-02-01 | bcc: update to 0.16.21
+2022-02-01 | opendbx - Native DBExpress driver v1.4.6 (mysql, postgres, odbc, sqli…
+2022-02-01 | update nodejs-14 to 14.19.0
```
Better phrase would be: opendbx-1.4.6 added

mebenn replied:
Slightly rephrased to: Add support for opendbx

AndWac commented on docs/release-notes/latest-changes.md:
```diff
@@ -28,6 +28,56 @@ All Rights Reserved. (Contributor contact(s):________________[Insert hyperlink/a
+2022-02-01 | chunga: update to 1.1.7
+2022-02-01 | pigz: update to 2.7
+2022-02-01 | disable some failing packages until we can provide fixes
+2022-02-01 | puri: update to 1.5.7
```
The entry ""disable some failing packages..." should also be removed.

mebenn replied:
Done

AndWac commented on docs/release-notes/latest-changes.md:
```diff
@@ -28,6 +28,56 @@ All Rights Reserved. (Contributor contact(s):________________[Insert hyperlink/a
+2022-01-31 | seahorse: fix for gnupg-2.3.4
+2022-01-31 | meld: fix build with meson > 0.60.0
+2022-02-01 |  php-812 - current version of php (based on openssl 1.1) includes fix…
+2022-01-31 | GNOME download server is https://download.gnome.org/
```
PHP-8.1.2 added

mebenn replied:
Add PHP-8.1.2

AndWac commented on docs/release-notes/latest-changes.md:
```diff
@@ -28,6 +28,56 @@ All Rights Reserved. (Contributor contact(s):________________[Insert hyperlink/a
+2022-01-30 | sshpass: update to 1.09
+2022-01-30 | expat: update to 2.4.4
+2022-01-30 | stress: update to 1.0.5 and add test results
+2022-01-29 | php-81 fix sed to cover all conditions to build correctly with openss…
```
"and add test results" should be removed.

mebenn replied:
Done

AndWac commented on docs/release-notes/latest-changes.md:
```diff
@@ -28,6 +28,56 @@ All Rights Reserved. (Contributor contact(s):________________[Insert hyperlink/a
+2022-01-30 | stress: update to 1.0.5 and add test results
+2022-01-29 | php-81 fix sed to cover all conditions to build correctly with openss…
+2022-01-01 | Upgrade of pygobject to 3.42.0 with python 3.7 and 3.9
+2021-07-30 | sudo: update to 1.9.7p2
```
Some of the entries are not in chronological order.

mebenn replied:
Damm, more manual work required :=(
I only strarted reading the GitHub REST API yesterday, and that is the order that the GET returns the enteries. I'll manually edit for now, but I'll probably have to do the sorting in the script. More importantly, I'm not sure if some closed PRs are missing!
...Done

JMadgwick reviewed (commented):
I notice there are some entries in the wrong place which are duplicates, such as the one for Firefox 60 which was already there.
I think the ultimate solution would be for this file to be updated automatically by GitHub Actions - I think this is probably possible, though it may be tricky to implement.

mebenn commented:
That list was generated by a python script using the GitHub REST API and as AndWac noticed, the sort order is incorrect, although I added a sort on 'closed_at' to my query, and the query reported no errors. I wasn't the only person to notice this, and this has been reported to the GitHub people, who have admitted to the abnormal behaviour : if you add a comment to a closed PR, it is this new date that is included in the sort, although the closed_at tag hadn't changed, and we are sorting on 'closed_at'.

So, I've implemented a sort in the script. So that is fixed.

Now, my original idea of retrieving closed pull requests was fraught with problems. The GitHub REST API looks really good on paper---which is why I initially decided to embark on automatically producing release notes; however, under the cover there are problems. GitHub only returns a certain amount of data; and the paging is really strange, and the date modifiers seem to modify paging through the data.

So I've got a new strategy implemented: I retrieve closed requests per day: no paging required as OI do not have significant amounts of PRs per day, so no paging required and  the script loops over the days. So you pass a date to the script and it retrieves all closed PRs from the current date back to the date supplied to the script. Moreover, the ouput is to a markup file; so, hopefully, nothing to do for the user. We should be able to start the script on Jenkins on every build, producing release notes on each build.

There is a problem though: GitHub limits the number of requests per hour. As long as the date is not too far back in the past, we should be ok. We can get around this, by extending the script so that a user can provide GitHub user-details + token; but this would take time, mainly due to handling such sensitive information securely.

I would also like to extend the release notes to include, for example, links to Redmine tickets, when available. Furthermore, not all PRs should be included in the release notes; but we could solve this by introducing a new tag to mark PRs for mention in the release notes.  

So, theoretically, it is possible to automate the process. Ultimately, the script should be a separate package, so that we can check for dependencies on  Python etc., but I'll commit and push to oi-docs as quick as possible so that you can have time to test, examine, etc

JMadgwick commented:
> There is a problem though: GitHub limits the number of requests per hour. As long as the date is not too far back in the past, we should be ok. We can get around this, by extending the script so that a user can provide GitHub user-details + token; but this would take time, mainly due to handling such sensitive information securely.

Providing a token can be done as part of a GitHub Actions CI job (we are no longer using Jenkins). The job could be triggered every week (or two) and update the change list with all changes to oi-userland since the change list was last updated. This can be done using [the schedule feature](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule).

I've not looked at the GitHub REST API for this just yet, but the comment about fetching changes on a per day basis for each request sounds very sensible. A script could take the most recent date in the change list and look up every day from then to the current date.
